libs/utils/energy_model.py

```diff
@@ -0,0 +1,572 @@
+            if not any(cpus_active[c] for c in pd.cpus):
+                if pd.parent:
+                    parent_state = find_deepest(pd.parent)
+                    if parent_state:
+                        return parent_state
+                return pd.idle_states[-1] if len(pd.idle_states) else None
+            return None
+
+        return [find_deepest(pd) for pd in self.cpu_pds]
+
+    def guess_idle_states(self, cpus_active):
+        """Pessimistically guess the idle states that each CPU may enter
+
+        If a CPU has any tasks it is estimated that it may only enter its
+        shallowest idle state in between task activations. If all the CPUs
+        within a power domain have no tasks, they will all be judged able to
+        enter that domain's deepest idle state. If any CPU in a domain has work,
+        no CPUs in that domain are assumed to enter any domain shared state.
+
+        e.g. Consider a system with
+
+        - two power domains PD0 and PD1
+
+        - 4 CPUs, with CPUs [0, 1] in PD0 and CPUs [2, 3] in PD1
+
+        - 4 idle states: "WFI", "cpu-sleep", "cluster-sleep-0" and "cluster-sleep-1"
+
+        Then here are some example inputs and outputs:
+
+        ::
+
+          # All CPUs idle:
+          [0, 0, 0, 0] -> ["cluster-sleep-0", "cluster-sleep-0",
+                           "cluster-sleep-0", "cluster-sleep-0"]
+
+          # All CPUs have work
+          [1, 1, 1, 1] -> ["WFI","WFI","WFI", "WFI"]
+
+          # One power domain active, the other idle
+          [0, 0, 1, 1] -> ["cluster-sleep-1", "cluster-sleep-1", "WFI","WFI"]
+
+          # One CPU active.
+          # Note that CPU 2 has no work but is assumed to never be able to enter
+          # any "cluster" state.
+          [0, 0, 0, 1] -> ["cluster-sleep-1", "cluster-sleep-1",
+                           "cpu-sleep","WFI"]
+
+        :param cpus_active: list where bool(cpus_active[N]) is False iff no
+                            tasks will run on CPU N.
+        :returns: List ``ret`` where ``ret[N]`` is the name of the estimated
+                  idle state that CPU N can enter during idle periods.
+
+        """
+        states = self._guess_idle_states(cpus_active)
+        return [s or c.idle_states.keys()[0]
+                for s, c in zip(states, self.cpu_nodes)]
+
+    def _guess_freqs(self, util_distrib):
+        overutilized = False
+        # Find what frequency each CPU would need if it was alone in its
+        # frequency domain
+        ideal_freqs = [0 for _ in self.cpus]
+        for node in self.cpu_nodes:
+            [cpu] = node.cpus
+            required_cap = util_distrib[cpu]
+
+            possible_freqs = [f for f, s in node.active_states.iteritems()
+                              if s.capacity >= required_cap]
+
+            if possible_freqs:
+                ideal_freqs[cpu] = min(possible_freqs)
+            else:
+                # CPU cannot provide required capacity, use max freq
```

Comment:
No, I don't agree that this can never happen - the `util_distrib` is not capped (and IIUC `util_avg` in a trace can be greater than the current capacity).